nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java

```diff
@@ -158,17 +191,43 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
             final MongoCursor<Document> cursor = it.iterator();
             try {
                 FlowFile flowFile = null;
-                while (cursor.hasNext()) {
-                    flowFile = session.create();
-                    flowFile = session.write(flowFile, new OutputStreamCallback() {
-                        @Override
-                        public void process(OutputStream out) throws IOException {
-                            IOUtils.write(cursor.next().toJson(), out);
+                if (context.getProperty(RESULTS_PER_FLOWFILE).isSet()) {
+                    int ceiling = context.getProperty(RESULTS_PER_FLOWFILE).asInteger();
+                    List<Document> batch = new ArrayList<>();
+
+                    while (cursor.hasNext()) {
+                        batch.add(cursor.next());
+                        if (batch.size() == ceiling) {
+                            try {
+                                getLogger().info("Writing batch...");
+                                String payload = buildBatch(batch);
+                                writeBatch(payload, context, session);
+                                batch = new ArrayList<>();
+                            } catch (IOException ex) {
+                                getLogger().error("Error building batch", ex);
+                            }
                         }
-                    });
-
-                    session.getProvenanceReporter().receive(flowFile, context.getProperty(URI).getValue());
-                    session.transfer(flowFile, REL_SUCCESS);
+                    }
+                    if (batch.size() > 0) {
+                        try {
+                            writeBatch(buildBatch(batch), context, session);
+                        } catch (IOException ex) {
+                            getLogger().error("Error sending remainder of batch", ex);
+                        }
+                    }
+                } else {
+                    while (cursor.hasNext()) {
+                        flowFile = session.create();
+                        flowFile = session.write(flowFile, new OutputStreamCallback() {
+                            @Override
+                            public void process(OutputStream out) throws IOException {
+                                IOUtils.write(cursor.next().toJson(), out);
+                            }
+                        });
+
```

Comment:
Could you also add the following?
````java
flowFile = session.putAttribute(flowFile, CoreAttributes.MIME_TYPE.key(), "application/json");
````
It's not mandatory but will improve UX when viewing flow file content in the UI.